File: Project 3/p3.f90
!MATH 96012 Project 3
! Alexander John Pinches CID:01201653
!This module contains four module variables and two subroutines;
!one of these routines must be developed for this assignment.
!Module variables--
! bm_l, bm_s0, bm_r0, bm_a: the parameters l, s0, r0, and a in the particle dynamics model
! numthreads: The number of threads that should be used in parallel regions within simulate2_omp
!
!Module routines---
! simulate2_f90: Simulate particle dynamics over m trials. Return: all x-y positions at final time
! and alpha at nt+1 times averaged across the m trials.
! simulate2_omp: Same input/output functionality as simulate2.f90 but parallelized with OpenMP

module bmotion
  implicit none
  integer :: numthreads
  real(kind=8) :: bm_l,bm_s0,bm_r0,bm_a
  real(kind=8), parameter :: pi = acos(-1.d0)
  complex(kind=8), parameter :: ii = complex(0.d0,1.d0)
contains

!Compute m particle dynamics simulations using the parameters,bm_l,bm_s0,bm_r0,bm_a.
!Input:
!m: number of simulations
!n: number of particles
!nt: number of time steps
!Output:
! x: x-positions at final time step for all m trials
! y: y-positions at final time step for all m trials
! alpha_ave: alignment parameter at each time step (including initial condition)
! averaged across the m trials
subroutine simulate2_f90(m,n,nt,x,y,alpha_ave)
  implicit none
  integer, intent(in) :: m,n,nt
  real(kind=8), dimension(m,n), intent(out) :: x,y
  real(kind=8), dimension(nt+1), intent(out) :: alpha_ave
  integer :: i1,j1,k1
  real(kind=8), dimension(m,n) :: nn !neighbors
  real(kind=8) :: r0sq !r0^2
  real(kind=8), dimension(m,n) :: phi_init,r_init,theta !used for initial conditions
  real(kind=8), dimension(m,n,n) :: dist2 !distance squared
  real(kind=8), allocatable, dimension(:,:,:) :: temp
  complex(kind=8), dimension(m,n) :: phase
  complex(kind=8), dimension(m,n,nt+1) :: exp_theta,AexpR




!---Set initial condition and initialize variables----
  allocate(temp(m,n,nt+1))
  call random_number(phi_init)
  call random_number(r_init)
  call random_number(theta)
  call random_number(temp)
  phi_init = phi_init*(2.d0*pi)
  r_init = sqrt(r_init)
  theta = theta*(2.d0*pi) !initial direction of motion
  x = r_init*cos(phi_init)+0.5d0*bm_l !initial positions
  y = r_init*sin(phi_init)+0.5d0*bm_l

  alpha_ave=0.d0
  r0sq = bm_r0*bm_r0
  exp_theta(:,:,1) = exp(ii*theta)
  AexpR = bm_a*exp(ii*temp*2.d0*pi) !noise term
  deallocate(temp)
  nn = 0

!----------------------------------------------
  !Time marching
  do i1 = 2,nt+1

    phase=0.d0
    dist2 = 0.d0

    !Compute distances
    do j1=1,n-1
      do k1 = j1+1,n
        dist2(:,j1,k1) = (x(:,j1)-x(:,k1))**2 + (y(:,j1)-y(:,k1))**2
        where (dist2(:,j1,k1)>r0sq)
          dist2(:,j1,k1)=0
        elsewhere
          dist2(:,j1,k1)=1
        end where
       dist2(:,k1,j1) =dist2(:,j1,k1)
      end do
    end do

    nn = sum(dist2,dim=3)+1

    !Update phase
    phase =  exp_theta(:,:,i1-1) +nn*AexpR(:,:,i1-1)
    do j1=1,m
      phase(j1,:) = phase(j1,:) + matmul(dist2(j1,:,:),exp_theta(j1,:,i1-1))
    end do

    !Update Theta
    theta = atan2(aimag(phase),real(phase))

    !Update X,Y
    x = x + bm_s0*cos(theta)
    y = y + bm_s0*sin(theta)

    x = modulo(x,bm_l)
    y = modulo(y,bm_l)

    exp_theta(:,:,i1) = exp(ii*theta)

  end do

  alpha_ave = (1.d0/dble(m*n))*sum(abs(sum(exp_theta,dim=2)),dim=1)


end subroutine simulate2_f90


!Same functionality as simulate2_f90, but parallelized with OpenMP
!Parallel regions should use numthreads threads.
!Compute m particle dynamics simulations using the parameters,bm_l,bm_s0,bm_r0,bm_a.
!Same functionality as simulate2_f90, but parallelized with OpenMP
!Parallel regions should use numthreads threads.
!Input:
!m: number of simulations
!n: number of particles
!nt: number of time steps
!Output:
! x: x-positions at final time step for all m trials
! y: y-positions at final time step for all m trials
! alpha_ave: alignment parameter at each time step (including initial condition)
! averaged across the m trials

! We assume n and m are large so any loop over either of these we should seek to parallise
! with everything inside the loop vectorised. Any sums can also be decomposed to loops. We 
! know we can't parallise over nt and we arent assuming this is large for the main loop so we parallise the do loops inside this. 
! For the nested loop in finding distances we make sure we arent having threads writing to the
! same or wrong parts of the array by making k1 private. Also using reduction for sums and calling 
! parallel do as few times as possible as forking takes time.
subroutine simulate2_omp(m,n,nt,x,y,alpha_ave)
  use omp_lib
  implicit none
  integer, intent(in) :: m,n,nt
  real(kind=8), dimension(m,n), intent(out) :: x,y
  real(kind=8), dimension(nt+1), intent(out) :: alpha_ave
  integer :: i1,j1,k1
  real(kind=8), dimension(m,n) :: phi_init,r_init,theta !used for initial conditions
  !Note: use allocatable arrays where possible to avoid OpenMP memory issues
  real(kind=8), allocatable, dimension(:,:,:) :: dist2 !distance squared
  real(kind=8), allocatable, dimension(:,:,:) :: temp
  complex(kind=8), allocatable,dimension(:,:) :: phase,temp2
  complex(kind=8), allocatable,dimension(:,:,:) :: exp_theta,AexpR
  real(kind=8), allocatable,dimension(:,:) :: nn
  real(kind=8) :: r0sq
!---Set initial condition and initialize variables (does not need to be parallelized)----
  call random_number(phi_init)
  call random_number(r_init)
  call random_number(theta)
  phi_init = phi_init*(2.d0*pi)
  r_init = sqrt(r_init)
  theta = theta*(2.d0*pi) !initial direction of motion
  x = r_init*cos(phi_init)+0.5d0*bm_l !initial positions
  y = r_init*sin(phi_init)+0.5d0*bm_l
!-------------------------------------------------
  

 !$ call omp_set_num_threads(numthreads)
  
  r0sq = bm_r0*bm_r0
  allocate(exp_theta(m,n,nt+1))
  !$OMP parallel do
  do i1=1,m
    exp_theta(i1,:,1)=exp(ii*theta(i1,:))
  end do
  !$OMP end parallel do
  allocate(temp(m,n,nt+1))
  call random_number(temp)
  allocate(AexpR(m,n,nt+1))
  !$OMP parallel do
  do i1=1,m
    AexpR(i1,:,:) = bm_a*exp(ii*temp(i1,:,:)*2.d0*pi)
  end do
  !$OMP end parallel do
  
  deallocate(temp)
  allocate(dist2(m,n,n))
  allocate(phase(m,n))
  allocate(nn(m,n))
!----------------------------------------------
  !Time marching
  
  do i1 = 2,nt+1
    !Compute distances
    
    dist2 = 0.d0
    !$OMP parallel do private(k1)
    do j1=1,n-1
      do k1 = j1+1,n
        dist2(:,k1,j1) = (x(:,j1)-x(:,k1))**2 + (y(:,j1)-y(:,k1))**2
        where (dist2(:,k1,j1)>r0sq)
          dist2(:,k1,j1)=0.d0
        elsewhere
          dist2(:,k1,j1)=1.d0
        end where
      end do
    end do
    !$OMP end parallel do
    
   
    
    
    nn = 1.d0
    !$OMP parallel do reduction(+:nn)
    do k1=1,n
      nn = nn + dist2(:,:,k1) 
    end do
    !$OMP end parallel do
    
    !Update phase
    
    phase=0.d0
    !$OMP parallel do 
    do j1=1,m
      phase(j1,:) = exp_theta(j1,:,i1-1) +nn(j1,:)*AexpR(j1,:,i1-1) + matmul(dist2(j1,:,:),exp_theta(j1,:,i1-1))
      theta(j1,:) = atan2(aimag(phase(j1,:)),real(phase(j1,:)))
      x(j1,:)=modulo(x(j1,:) + bm_s0*cos(theta(j1,:)),bm_l)
      y(j1,:)=modulo(y(j1,:) + bm_s0*sin(theta(j1,:)),bm_l)
      exp_theta(j1,:,i1) = exp(ii*theta(j1,:))
      
    end do
    !$OMP end parallel do
    
    
    
  end do
  ! deallocate
  deallocate(nn)
  deallocate(phase)
  deallocate(AexpR)
  ! calculate alpha
  allocate(temp2(m,nt+1))
  temp2=0.d0
  !$OMP parallel do reduction(+:temp2)
  do i1=1,n
    temp2 = temp2 + exp_theta(:,i1,:)
  end do
  !$OMP end parallel do 
  deallocate(exp_theta)
  alpha_ave=0.d0
  !$OMP parallel do reduction(+:alpha_ave)
  do i1=1,m
    alpha_ave = alpha_ave + (1.d0/dble(m*n))*abs(temp2(i1,:))
  end do
  !$OMP end parallel do 
  

end subroutine simulate2_omp

subroutine allPositions(n,nt,x,y)
  use omp_lib
  implicit none
  integer :: n,nt
  real(kind=8), dimension(n,nt+1), intent(out) :: x,y
  integer :: i1,j1,k1
  real(kind=8), allocatable, dimension(:) :: phi_init,r_init,theta !used for initial conditions
  !Note: use allocatable arrays where possible to avoid OpenMP memory issues
  real(kind=8), allocatable, dimension(:,:) :: dist2,temp !distance squared
  complex(kind=8), allocatable,dimension(:) :: phase
  complex(kind=8), allocatable,dimension(:,:) :: exp_theta,AexpR
  real(kind=8), allocatable,dimension(:) :: nn
  real(kind=8) :: r0sq
!---Set initial condition and initialize variables (does not need to be parallelized)----
  allocate(phi_init(n))
  allocate(r_init(n))
  allocate(theta(n))
  call random_number(phi_init)
  call random_number(r_init)
  call random_number(theta)
  phi_init = phi_init*(2.d0*pi)
  r_init = sqrt(r_init)
  theta = theta*(2.d0*pi) !initial direction of motion
  x(:,1) = r_init*cos(phi_init)+0.5d0*bm_l !initial positions
  y(:,1) = r_init*sin(phi_init)+0.5d0*bm_l
  deallocate(r_init)
  deallocate(phi_init)
!-------------------------------------------------
  !$ call omp_set_num_threads(numthreads)
  
  r0sq = bm_r0*bm_r0
  allocate(exp_theta(n,nt+1))
  !$OMP parallel do
  do i1=1,n
    exp_theta(i1,1)=exp(ii*theta(i1))
  end do
  !$OMP end parallel do
  allocate(temp(n,nt+1))
  call random_number(temp)
  allocate(AexpR(n,nt+1))

  !$OMP parallel do
  do i1=1,n
    AexpR(i1,:) = bm_a*exp(ii*temp(i1,:)*2.d0*pi)
  end do
  !$OMP end parallel do

  deallocate(temp)
  allocate(dist2(n,n))
  allocate(phase(n))
  allocate(nn(n))
!----------------------------------------------
  !Time marching
  
  do i1 = 2,nt+1
    !Compute distances
    
    dist2 = 0.d0
    !$OMP parallel do private(k1)
    do j1=1,n-1
      do k1 = j1+1,n
        dist2(k1,j1) = (x(j1,i1-1)-x(k1,i1-1))**2 + (y(j1,i1-1)-y(k1,i1-1))**2
        if (dist2(k1,j1)>r0sq) then 
          dist2(k1,j1)=0.d0
        else 
          dist2(k1,j1)=1.d0
        end if
      end do
    end do
    !$OMP end parallel do
    

    
    
    nn = 1.d0
    !$OMP parallel do reduction(+:nn)
    do k1=1,n
      nn = nn + dist2(:,k1) 
    end do
    !$OMP end parallel do
    
    !Update phase

    phase=0.d0
    !$OMP parallel do
    do j1=1,n
      phase(j1) = exp_theta(j1,i1-1) +nn(j1)*AexpR(j1,i1-1)
    end do
    !$OMP end parallel do

    phase = phase + matmul(dist2,exp_theta(:,i1-1))
    ! calculate new positions
    !$OMP parallel do
    do j1=1,n
      theta(j1) = atan2(aimag(phase(j1)),real(phase(j1)))
      x(j1,i1)=modulo(x(j1,i1-1) + bm_s0*cos(theta(j1)),bm_l)
      y(j1,i1)=modulo(y(j1,i1-1) + bm_s0*sin(theta(j1)),bm_l)
      exp_theta(j1,i1) = exp(ii*theta(j1))
    end do
    !$OMP end parallel do
    
    
    
    
  end do
 
end subroutine allPositions

subroutine correlation(a,b,m,c)
  use omp_lib
  implicit none
  integer :: a,b,m,nt,i1
  real(kind=8), dimension(81), intent(out) :: c
  real(kind=8), allocatable, dimension(:) :: alpha_ave
  real(kind=8) :: last
  real(kind=8), allocatable,dimension(:,:) :: x,y

  nt= a + b + 80
  allocate(alpha_ave(nt+1))
  allocate(x(m,400))
  allocate(y(m,400))
  ! get alphas using simulate2_omp
  call simulate2_omp(m=m,n=400,nt=nt,x=x,y=y,alpha_ave=alpha_ave)
  last=0.d0
  !$ call omp_set_num_threads(numthreads)
  ! calculate last term
  !$OMP parallel do reduction(+:last)
  do i1=a,b
    last = last + (alpha_ave(i1)/dble(b-a))**2
  end do
  !$OMP end parallel do
  ! calculate c for each tau
  !$OMP parallel do
  do i1 = 1,81
    c(i1) = SUM(alpha_ave(i1+a-1:b+i1-1)*alpha_ave(a:b))/dble(b-a) + last
  end do
  !$OMP end parallel do
end subroutine correlation
end module bmotion

File: Project 3/p3.py
"""MATH 96012 Project 3
Alexander John Pinches CID:01201653
Contains four functions:
    simulate2: Simulate bacterial dynamics over m trials. Return: all positions at final time
        and alpha at nt+1 times averaged across the m trials.
    performance: To be completed -- analyze and assess performance of python, fortran, and fortran+openmp simulation codes
    correlation: To be completed -- compute and analyze correlation function, C(tau)
    visualize: To be completed -- generate animation illustrating "non-trivial" particle dynamics
"""
import numpy as np
import matplotlib.pyplot as plt
from m1 import bmotion as bm #assumes that p3_dev.f90 has been compiled with: f2py --f90flags='-fopenmp' -c p3_dev.f90 -m m1 -lgomp
import scipy.spatial.distance as scd
#May also use scipy and time modules as needed
import time
import matplotlib.animation as animation

plt.style.use("ggplot")

def simulate2(M=10,N=64,L=8,s0=0.2,r0=1,A=0,Nt=100):
    """Simulate bacterial colony dynamics
    Input:
    M: Number of simulations
    N: number of particles
    L: length of side of square domain
    s0: speed of particles
    r0: particles within distance r0 of particle i influence direction of motion
    of particle i
    A: amplitude of noise
    Nt: number of time steps

    Output:
    X,Y: position of all N particles at Nt+1 times
    alpha: alignment parameter at Nt+1 times averaged across M simulation

    Do not modify input or return statement without instructor's permission.

    Add brief description of approach of differences from simulate1 here:
    This code carries out M simulations at a time with partial vectorization
    across the M samples.
    """
    #Set initial condition
    phi_init = np.random.rand(M,N)*(2*np.pi)
    r_init = np.sqrt(np.random.rand(M,N))
    Xinit,Yinit = r_init*np.cos(phi_init),r_init*np.sin(phi_init)
    Xinit+=L/2
    Yinit+=L/2
    #---------------------

    #Initialize variables
    P = np.zeros((M,N,2)) #positions
    P[:,:,0],P[:,:,1] = Xinit,Yinit
    alpha = np.zeros((M,Nt+1)) #alignment parameter
    S = np.zeros((M,N),dtype=complex) #phases
    T = np.random.rand(M,N)*(2*np.pi) #direction of motion
    n = np.zeros((M,N)) #number of neighbors
    E = np.zeros((M,N,Nt+1),dtype=complex)
    d = np.zeros((M,N,N))
    dtemp = np.zeros((M,N*(N-1)//2))
    AexpR = np.random.rand(M,N,Nt)*(2*np.pi)
    AexpR = A*np.exp(1j*AexpR)

    r0sq = r0**2
    E[:,:,0] = np.exp(1j*T)

    #Time marching-----------
    for i in range(Nt):
        for j in range(M):
            dtemp[j,:] = scd.pdist(P[j,:,:],metric='sqeuclidean')

        dtemp2 = dtemp<=r0sq
        for j in range(M):
            d[j,:,:] = scd.squareform(dtemp2[j,:])
        n = d.sum(axis=2) + 1
        S = E[:,:,i] + n*AexpR[:,:,i]

        for j in range(M):
            S[j,:] += d[j,:,:].dot(E[j,:,i])

        T = np.angle(S)

        #Update X,Y
        P[:,:,0] = P[:,:,0] + s0*np.cos(T)
        P[:,:,1] = P[:,:,1] + s0*np.sin(T)

        #Enforce periodic boundary conditions
        P = P%L

        E[:,:,i+1] = np.exp(1j*T)
    #----------------------

    #Compute order parameter
    alpha = (1/(N*M))*np.sum(np.abs(E.sum(axis=1)),axis=0)

    return P[:,:,0],P[:,:,1],alpha


def performance(input_p=(None),display=False):
    """Assess performance of simulate2, simulate2_f90, and simulate2_omp
    Modify the contents of the tuple, input, as needed
    When display is True, figures equivalent to those
    you are submitting should be displayed
    input_p=(s0,A,L,r0,num_threads,M,N,Nt)
    int :: s0,A,L,r0,num_threads
    list of ints :: M,N,Nt
    """
    # set parameters in fortran module
    bm.bm_s0=input_p[0]
    bm.bm_a=input_p[1]
    bm.bm_l=input_p[2]
    bm.bm_r0=input_p[3]
    bm.numthreads=input_p[4]
    # extract values to time for
    m = input_p[5]
    n = input_p[6]
    nt = input_p[7]
    # create dictionaries to store results
    mTimes={"F90+OMP":[None]*len(m),"F90":[None]*len(m),"PYTHON":[None]*len(m)}
    nTimes={"F90+OMP":[None]*len(n),"F90":[None]*len(n),"PYTHON":[None]*len(n)}
    ntTimes={"F90+OMP":[None]*len(n),"F90":[None]*len(nt),"PYTHON":[None]*len(nt)}
    # time for m
    for v,i in enumerate(m): 
        print(f"Timing for M={i}...")   
        mTimes["F90+OMP"][v]=timer(bm.simulate2_omp,m=i,n=100,nt=100)
        mTimes["F90"][v]=timer(bm.simulate2_f90,m=i,n=100,nt=100)
        mTimes["PYTHON"][v]=timer(simulate2,M=i,N=100,Nt=100,A=input_p[1])
    # time for n
    for v,i in enumerate(n):  
        print(f"Timing for N={i}...")  
        nTimes["F90+OMP"][v]=timer(bm.simulate2_omp,n=i,m=100,nt=100)
        nTimes["F90"][v]=timer(bm.simulate2_f90,n=i,m=100,nt=100)
        nTimes["PYTHON"][v]=timer(simulate2,N=i,M=100,Nt=100,A=input_p[1]) 
    # time for nt
    for v,i in enumerate(nt):
        print(f"Timing for Nt={i}...")    
        ntTimes["F90+OMP"][v]=timer(bm.simulate2_omp,n=100,m=100,nt=i)
        ntTimes["F90"][v]=timer(bm.simulate2_f90,n=100,m=100,nt=i)
        ntTimes["PYTHON"][v]=timer(simulate2,N=100,M=100,Nt=i,A=input_p[1])   
    # increasing m figure       
    plt.figure()
    plt.plot(m,mTimes["F90+OMP"],label="F90+OMP")
    plt.plot(m,mTimes["F90"],label="F90")
    plt.plot(m,mTimes["PYTHON"],label="PYTHON")
    plt.legend()
    plt.title("Run time for increasing M and N=100 and Nt=100")
    plt.xlabel("M")
    plt.ylabel("Time (s)")
    plt.savefig("m_time_comparison.png")
    # increasing N figure
    plt.figure()
    plt.plot(n,nTimes["F90+OMP"],label="F90+OMP")
    plt.plot(n,nTimes["F90"],label="F90")
    plt.plot(n,nTimes["PYTHON"],label="PYTHON")
    plt.legend()
    plt.title("Run time for increasing N and M=100 and Nt=100")
    plt.xlabel("N")
    plt.ylabel("Time (s)")
    plt.savefig("n_time_comparison.png")
    # increasin nt figure
    plt.figure()
    plt.plot(nt,ntTimes["F90+OMP"],label="F90+OMP")
    plt.plot(nt,ntTimes["F90"],label="F90")
    plt.plot(nt,ntTimes["PYTHON"],label="PYTHON")
    plt.legend()
    plt.title("Run time for increasing Nt and N=100 and M=100")
    plt.xlabel("Nt")
    plt.ylabel("Time (s)")
    plt.savefig("nt_time_comparison.png")
    
    if display: # if display==True show
        plt.show()
    
    return mTimes,nTimes,ntTimes #Modify as needed

def timer(func,repeats=1,**kwargs):
    """
    func = function to time
    repeats = repeats to average over
    **kwargs = kwargs to be passed to func
    """
    l=[None]*repeats # store results
    for i in range(repeats): # time for repeats
        start = time.time()
        func(**kwargs)
        end = time.time()
        l[i] = end - start
    return np.mean(l) # reaturn mean time
    
    
def correlation(input_c=(None),display=False):
    """Compute and analyze temporal correlation function, C(tau)
    Modify the contents of the tuple, input, as needed
    When display is True, figures equivalent to those
    you are submitting should be displayed
    input_c=(s0,A,L,r0,num_threads,a,M)
    int :: s0,A,L,r0,num_threads
    """
    # set fortran module variables
    bm.bm_s0=input_c[0]
    bm.bm_a=input_c[1]
    bm.bm_l=input_c[2]
    bm.bm_r0=input_c[3]
    bm.numthreads=input_c[4]
    # set a,b and m
    a= input_c[5]
    b = 200 +a
    m = input_c[6]
    # compute correlation for tau 0 to 80
    c=bm.correlation(a,b,m)
    # plot
    x=np.linspace(0,80,81)
    plt.figure()
    plt.plot(x,c)
    plt.title(r"$C(\tau)$")
    plt.xlabel(r'$ \tau $')
    plt.ylabel(r"$C(\tau)$")
    plt.savefig("corr.png")
    if display: # if display==True show
        plt.show()
    return c #Modify as needed

    
def updatefig(i,x,y,line):
    """Updates figure each time function is called
    and returns new figure 'axes'
    """
    line.set_ydata(y[:,i])
    line.set_xdata(x[:,i])
    return line,    


def visualize(n=128,nt=500,s0=0.1,A=0.65,l=4,r0=1):
    """Generate an animation illustrating particle dynamics
    """
    # set variables in fortran
    bm.bm_s0=s0
    bm.bm_a=A
    bm.bm_l=l
    bm.bm_r0=r0
    bm.numthreads=2
    x,y = bm.allpositions(n=n,nt=nt) # faster to generate all values and index
    fig, ax=plt.subplots()
    # create plot
    line,=ax.plot(x[:,0],y[:,0],linewidth=0,markersize=8,marker='o')
    ax.set_xlabel("X")
    ax.set_ylabel("Y")
    ax.set_title(f"Animation for N={n}, s0={s0}, A={A}, l={l}, r0={r0}")
    # set limits to l
    ax.set_xlim(0,l)
    ax.set_ylim(0,l)
    # animate
    ani = animation.FuncAnimation(fig, updatefig,frames=30, repeat=False,blit=True,fargs=(x,y,line))
    ani.save("p3movie.mp4",writer="ffmpeg")
    return x,y #Modify as needed


if __name__ == '__main__':
    #Modify the code here so that it calls performance analyze and
    # generates the figures that you are submitting with your code
    input_p = (0.2,0.64,8,1,2,[100,200,300,400,500],[100,200,300,400,500],[100,250,500,750,1000])
    output_p = performance(input_p) #modify as needed

    input_c = (0.1,0.625,16,1,2,100,200)
    output_c = correlation(input_c)
    
    visualize()
